Add git-webkit conflict command.

bravc opened the pull request:
#### fe858963a7c6088e8c9307e7aa5224176bad9641
<pre>
Add git-webkit conflict command.

<a href="https://rdar.apple.com/125747569">rdar://125747569</a> (git-safari/webkit conflict command for engineers)

Reviewed by Jonathan Bedard.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py: Added.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/conflict_unittest.py: Added.

Canonical link: <a href="https://commits.webkit.org/279720@main">https://commits.webkit.org/279720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9dc8a65e0ee4f24e6361fddfc6cf6d12b8ed7629

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54299 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/33705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6852 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/57580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/5030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/56603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/41226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/4975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/57580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/5030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56395 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/41226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/47015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/57580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/54131 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/41226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/4344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/3176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/41226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/4554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/59171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/4975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/59171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/54296 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/30680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47117 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/59171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31653 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8044 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/30462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -0,0 +1,84 @@
+    help = "Given the representative Radar ID of a conflicting merge, checkout the branch and setup the integration remote if needed."
+
+    INTEGRATION_REMOTE_NAME = "integration"
+    INTEGRATION_REMOTE_URL = "git@github.com:apple/WebKit-integration.git"
```
I think we should do something a bit more complex, and generalized, here.

`git-webkit` defines a set of canonical remotes, accessible from the `repository.source_remotes()`. I think we should assume that an integration remote will always be `<source-remote>/<project>-integration`. We should check each source remote, there won't be many, and this will allow us to support Integration remotes on any source remote for any project.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -0,0 +1,84 @@
+        )
+
+    @classmethod
+    def ensure_integration_remote_setup(cls, repository):
```
We want to delegate this to `repository.checkout()`, in particular, we don't have any guarantee that contributors have ssh keys configured, as we support pure https checkouts. I would consult the `checkout` program, I think you're looking to do something similar here.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -0,0 +1,84 @@
+            return 1
+        if not repository.path:
+            sys.stderr.write('Cannot checkout conflict on remote repository\n')
+            return 1
```
There should be another check, right? As this logic only applies to a `git` checkout?

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -0,0 +1,84 @@
+            sys.stderr.write('Failed to add {} as remote.'.format(cls.INTEGRATION_REMOTE_URL))
+            return 1
+
+        run([repository.executable(), 'fetch', cls.INTEGRATION_REMOTE_NAME], cwd=repository.root_path, capture_output=False)
```
I think before we fetch a specific remote, we should search for a PR with the expected branch name. This might end up making this a much smaller program, actually, because you completely delegate the rest of the checkout operation to `repository.checkout` once you know the remote and branch of the PR in question. This is probably going to look similar to what `git-webkit pr` does to search for an existing PR matching the provided branch name.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -0,0 +1,67 @@
+            if conflict_pr:
+                pr_target = conflict_pr._metadata['full_name']
+                target, _ = pr_target.split('/')
+                return repository.checkout('{}:{}'.format(target, conflict_pr.head))
```
This is essentially the same bug I fixed for buildbot in https://commits.webkit.org/279577@main. This should actually be `{}-integration`, specifically, although generally, when you split the PR target, you want it to be `{}-{<trailing descriptor>}`, I think. I will be exploring a fix for this in `git-webkit checkout` shortly, which I think has the same problem.

JonWBedard replied:
Actually, testing with `git-webkit checkout` demonstrates that we don't actually need the `split`, if we feed the `full_name` directly as the target, `repository.checkout` will do the right thing.

Commit:
Add git-webkit conflict command.

rdar://125747569 (git-safari/webkit conflict command for engineers)

Reviewed by Jonathan Bedard.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py: Added.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/conflict_unittest.py: Added.

Canonical link: https://commits.webkit.org/279720@main

fujii commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
```diff
@@ -95,6 +96,8 @@ def main(
         Clean, Clone, Find, Info, Land, Log, Pull,
         PullRequest, Revert, Review, Setup, InstallGitLFS,
+        Clean, Clone, Conflict, Find, Info, Land, Log, Pull,
+        PullRequest, Revert, Setup, InstallGitLFS,
```
This change broke buildbots.
Bug 275137 – REGRESSION(279720@main): argparse.ArgumentError: argument {help,blame,branch,canonicalize,checkout,clean,clone,find,list,info,land,log,pull,up,update,pull-request,pr,pfr,upload,revert,review,setup,install-git-lfs}: conflicting subparser: clean
https://bugs.webkit.org/show_bug.cgi?id=275137
https://github.com/WebKit/WebKit/pull/29526
